Add new string walk benchmarks

rctcwyvrn opened the pull request:
Add new string walk benchmarks for https://github.com/apple/swift/pull/59182 

The string walk benchmarks were previously set as `.skip`, I reenabled them for now and we can evaluate if we want to keep them that way

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

rctcwyvrn committed:
Add benchmarks for the walking a string by index

rctcwyvrn committed:
Remove .skip from string walk benchmarks

rctcwyvrn commented:
@swift-ci please benchmark

lorentey commented:
@swift-ci benchmark

rctcwyvrn committed:
Add a setup phase to warm up grapheme breaking

This may not fix things, just checking to see if the setup time warnings are from the grapheme breaking system warming up

rctcwyvrn commented:
@swift-ci please benchmark

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn committed:
Increase workload in string benchmarks

rctcwyvrn commented:
@swift-ci please benchmark

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn committed:
Make iteration benchmark strings longer

Instead of having many iterations

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn committed:
Rename benchmarks and adjust workkloads again

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn committed:
Fix more names and do more benchmark tuning

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn commented:
@swift-ci smoke test

lorentey reviewed (approved):
LGTM 🎉

lorentey commented:
Ah, we may want to have `StringWalk` continue to use its original implementation (but keep the new & better one for `StringWalk.ascii.scalars`). That's the only benchmark here used to be enabled before this PR, and it's nice to preserve continuity with older results.

```
------- Performance (x86_64): -O -------
...
StringWalk                                                  640     1160    +81.2%    **0.55x**

------- Code size: -O -------

REGRESSION     OLD     NEW     DELTA    RATIO  
StringWalk.o   36623   54785   +49.6%   **0.67x**
```

rctcwyvrn committed:
Merge branch 'main' into lily-string-benchmarks

rctcwyvrn committed:
more string benchmark stuff

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn commented:
I switched `StringWalk` to run the old code and did some other small stuff (removed legacy factor on the new benchmarks, moved string initialization to a function)

Maybe just maybe the string initializations were causing the setup overhead and this will fix it

rctcwyvrn committed:
Remove the extra scaling for the legacy factor

rctcwyvrn commented:
@swift-ci benchmark

rctcwyvrn commented:
This is probably about as good as we're going to get this

1. The doctor is mostly happy with the new string walk benchmarks, still just some setup overheads appearing at random
2. Had to modify some of the base strings to make things work better, but that ended up causing regressions/improvements in the char iteration/indexing benchmarks anyway. `punctuated` and `punctuatedJapanese` have major "regressions" due to adjusting the string to be longer. Even still the string walk benchmarks for those strings are still too short 

Honestly maybe trying to use these gyb'd string walk benchmarks in the first place was a mistake, trying to get them to work well while also not breaking the old benchmarks was a major pain and I ended up with the worst of both worlds anyway. The new benchmarks are not ideal and the old benchmarks are still modified.

Maybe it would be best to just throw this out and just make a set of string walk benchmarks from scratch, not attached to these existing ones at all. 

Either that or we can be fine with somewhat messing up the char iteration/indexing benchmarks and just go ahead and merge it
